### PR TITLE
Fix numerical power parsing for FITS-compatible units

### LINKED ISSUE #7681: FITS-standard unit parsing fails on some types of exponents
Why don't these work:
```python
from astropy.units import Unit
Unit('10**17 erg/(cm2 s Angstrom)', format='fits')
Unit('10^17 erg/(cm2 s Angstrom)', format='fits')
```
When these all do:
```python
from astropy.units import Unit
Unit('10+17 erg/(cm2 s Angstrom)', format='fits')
Unit('10**-17 erg/(cm2 s Angstrom)', format='fits')
Unit('10^-17 erg/(cm2 s Angstrom)', format='fits')
Unit('10-17 erg/(cm2 s Angstrom)', format='fits')
```

The non-working versions give *e.g.*:
```
ValueError: '10^17 erg/(cm2 s Angstrom)' did not parse as fits unit: Numeric factor not supported by FITS
```
which is not how I would interpret the [FITS standard](https://fits.gsfc.nasa.gov/standard30/fits_standard30aa.pdf).

Tested on 2.0.7 and 3.0.3

### PR DESCRIPTION
**Pull request by weaverba137**
This PR fixes #7681.

DO NOT MERGE YET.

I recalled that although @mhvk and I never did come up with a solution to this problem, I did refactor the way the units were tested, and that in itself is a useful starting point for testing and solving this problem.

### CODE REVIEW

**Comment by weaverba137** [comment]
This set of unit parser directives seems to work successfully with all cases.  However, it seems like this set should be smaller.

```python
            '''
            factor_fits : UINT power OPEN_PAREN signed_int CLOSE_PAREN
                        | UINT power OPEN_PAREN UINT CLOSE_PAREN
                        | UINT power signed_int
                        | UINT power UINT
                        | UINT SIGN UINT
                        | UINT OPEN_PAREN signed_int CLOSE_PAREN
            '''
```

**Comment by weaverba137** [comment]
Could someone please assign @mhvk  to review this?

**Comment by mhvk** [comment]
@weaverba137 - this looks very promising! I hope to look in more detail later (a bit swamped right now).

**Comment by weaverba137** [comment]
The current Travis failures appear to be completely unrelated to this PR.

**Comment by weaverba137** [comment]
Also, please review the CHANGES.rst file.  I'm not sure if the entries are in the right places any longer.

**Comment by weaverba137** [comment]
@mhvk, I added those other PRs to CHANGES.rst because I never received credit for that PR in that file.  For some reason, at the time #7658 was merged, it was not added to CHANGES.rst.  What do you think is the appropriate way to address this?

Otherwise, I'll get this added to 2.0.13 and squash.

**Comment by mhvk** [comment]
OK, didn't know that. A separate PR is probably better, since the backporting is different...

**Comment by weaverba137** [comment]
OK, fair enough.  I'll get working on the squash then.

**Comment by bsipocz** [comment]
@weaverba137 - Not all PRs contain changes that end up in the changelog. Please don't add unrelated entries in PRs as it  will mess with our release scripts.

**Comment by mhvk** [comment]
@weaverba137 - ah, I now looked at #7658 and that indeed is not something users really have to worry about, which is why it didn't need a change-log entry. Of course, this does not mean the contribution doesn't get recognized! That accounting is more simply done via commit authorship - we definitely don't want to ignore contributions and improvements to tests, error messages and documentation!

**Comment by weaverba137** [comment]
@bsipocz, that's fine.  I am not particularly concerned about it anymore; it was a while ago.  I'll get things cleaned up.

File: CHANGES.rst [hunk 1]
```diff
@@ -381,6 +381,8 @@ astropy.uncertainty
 astropy.units
 ^^^^^^^^^^^^^
 
+- Fix parsing of numerical powers in FITS-compatible units. [#8251]
```
Comment (bsipocz) [inline CHANGES.rst]: changelog only needed to be added to the section a PR is milestoned to.

**Comment by weaverba137** [comment]
There's the git equivalent of pages of algebra under the hood, but commits are squashed, and the change log entries are in the right places with no extraneous entries.

File: CHANGES.rst (hunk 1 above)
Comment (weaverba137) [reply CHANGES.rst]: Ah, I misunderstood.  It sounded like @mhvk wanted me to add two entries.

File: astropy/units/format/generic_parsetab.py [hunk 2]
```diff
@@ -1,24 +1,13 @@
-# -*- coding: utf-8 -*-
```
Comment (mhvk) [inline astropy/units/format/generic_parsetab.py]: One last very annoying thing: could you put the comment on top back?

Comment (weaverba137) [reply astropy/units/format/generic_parsetab.py]: Will do.

**Comment by weaverba137** [comment]
Comment header restored.

**Comment by bsipocz** [comment]
Given the version differences for the generated parsetab between LTS and master, I'm not backporting this.

**Pull request merged**